libraries/ESP8266WiFi/src/ESP8266WiFiGeneric.cpp

```diff
@@ -386,22 +394,36 @@ bool ESP8266WiFiGenericClass::getPersistent(){
  * set new mode
  * @param m WiFiMode_t
  */
-bool ESP8266WiFiGenericClass::mode(WiFiMode_t m) {
+bool ESP8266WiFiGenericClass::mode(WiFiMode_t m, WiFiState* state) {
+    if (m == WIFI_SHUTDOWN) {
+        return shutdown(0, state);
+    }
+    else if (m == WIFI_RESUME) {
+        return resumeFromShutdown(state);
+    }
+    else if (m & ~(WIFI_STA | WIFI_AP))
+        // any other bits than legacy disallowed
+        return false;
+
+    // m is now WIFI_STA, WIFI_AP or WIFI_AP_STA
+
+    if (wifi_fpm_get_sleep_type() != NONE_SLEEP_T) {
+        // wifi may have been put asleep by ESP8266WiFiGenericClass::preinitWiFiOff
+        wifi_fpm_do_wakeup();
+        wifi_fpm_close();
+    }
+
     if(_persistent){
         if(wifi_get_opmode() == (uint8) m && wifi_get_opmode_default() == (uint8) m){
+            saveState(state, m, _persistent);

```

Comment:
I don't get it.
If the mode is already the same as the one you want to set, then save the state?

Reply:
This is right.
saveState should be used only with shutdown (write all structure members) or resume mode (real). Fixed.